File: src/solutions/2024/19/ts/main.ts
import { argv } from "node:process";
import type { SolutionOptions } from "src/options.ts";
import { run } from "src/run.ts";

const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

function recurse(
   design: string,
   patterns: string[],
   memo: Map<string, number>,
): number {
   if (memo.has(design)) {
      return memo.get(design)!;
   }
   if (design.length === 0) {
      return 1;
   }
   let total = 0;
   for (const pattern of patterns) {
      if (design.startsWith(pattern)) {
         total += recurse(design.slice(pattern.length), patterns, memo);
      }
      memo.set(design, total);
   }
   return total;
}

function fastP1(
   design: string,
   patterns: string[],
   visited: Set<string>,
): boolean {
   if (design.length === 0) {
      return true;
   }
   if (visited.has(design)) {
      return false;
   }
   visited.add(design);
   for (const pattern of patterns) {
      if (design.startsWith(pattern)) {
         if (fastP1(design.slice(pattern.length), patterns, visited)) {
            return true;
         }
      }
   }
   return false;
}

function solve(input: string, p2: boolean): string {
   const lines = input.split("\n\n");
   const patterns = lines[0].split(", ");
   let total = 0;
   const memo = new Map<string, number>();
   for (const design of lines[1].split("\n")) {
      if (p2) {
         total += recurse(design, patterns, memo);
      }
      else if (fastP1(design, patterns, new Set())) {
         total++;
      }
   }
   return total.toString();
}

function part1(input: string, _isTest: boolean): string {
   return solve(input, false);
}

function part2(input: string, _isTest: boolean): string {
   return solve(input, true);
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}

File: src/solutions/2024/19/go/main.go
package main

import (
	core "advent/of/go/src/langs/go"
	"os"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

func recurse(design string, patterns []string, memo map[string]int) (ways int) {
	if ways, ok := memo[design]; ok {
		return ways
	}
	defer func() { memo[design] = ways }()

	if design == "" {
		return 1
	}

	for _, pattern := range patterns {
		if strings.HasPrefix(design, pattern) {
			ways += recurse(design[len(pattern):], patterns, memo)
		}
	}

	return ways
}

func solve(input string, p2 bool) string {
	lines := strings.Split(input, "\n\n")
	patterns := strings.Split(lines[0], ", ")

	memo := map[string]int{}
	total := 0
	for _, design := range strings.Split(lines[1], "\n") {
		combinations := recurse(design, patterns, memo)
		if combinations > 0 {
			if p2 {
				total += combinations
			} else {
				total++
			}
		}
	}

	return strconv.FormatInt(int64(total), 10)
}

func part1(input string, isTest bool) string {
	return solve(input, false)
}

func part2(input string, isTest bool) string {
	return solve(input, true)
}
